### PR TITLE
Task homework10

### PR DESCRIPTION
**Pull request by ritabunets**
Thanks in advance :)

### CODE REVIEW

File: Homework10/GenericArray.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    public class GenericArray<T> where T : Human
+    {
+        private T[] _array;
+
+        public GenericArray(int size)
+        {
+            _array = new T[size + 1];
+        }
+
+        public int GetAraryLength() => _array.Length;
```
Comment (yauhenists) [inline Homework10/GenericArray.cs]: If method is used only in this class make it private and move after all public and protected methods

File: Homework10/GenericArray.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+        public void AddElement(int index, T element)
+        {
+            _array[index] = element;
+        }
+
+        public T GetElement(int index) => _array[index];
+
+        public void RemoveElement(ref T[] _array, int index)
```
Comment (yauhenists) [inline Homework10/GenericArray.cs]: Naming rule for parameters - starts with lower case letter

File: Homework10/GenericArray.cs [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        {
+            T[] tmp = new T[_array.Length - 1];
+            for (int i = 0; i < index; i++)
```
Comment (yauhenists) [inline Homework10/GenericArray.cs]: Yes, your solution with deleting element works, but you can avoid using 2 loops. Approach with one loop:

            T[] temp2 = new T[_array.Length - 1];
            var j = 0;
            for (int i = 0; i < temp2.Length; i++)
            {
                if (i == index)
                {
                    continue;
                }

                temp2[j] = _array[i];
                j++;
            }

File: Homework10/GenericArray.cs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+                tmp[i] = _array[i];
+            }
+            for (int i = index; i < tmp.Length; i++)
+            {
+                tmp[i] = _array[i + 1];
+            }
+            _array = tmp;
+        }
+
+        public static void ToString(GenericArray<T> _array)
```
Comment (yauhenists) [inline Homework10/GenericArray.cs]: Please rewrite this method - make it not static and override this method

File: Homework10/GenericArray.cs [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        private T[] _array;
+
+        public GenericArray(int size)
+        {
+            _array = new T[size + 1];
+        }
+
+        public int GetAraryLength() => _array.Length;
+
+        public void AddElement(int index, T element)
```
Comment (yauhenists) [inline Homework10/GenericArray.cs]: Please make this method universal. It can throw IndexOutOfRangeException

File: Homework10/Program.cs [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+            GenerateElements(2, woman);
+            GenericArray<Woman>.ToString(woman);
+
+            GenerateElements(1, man);
+            GenericArray<Man>.ToString(man);
+        }
+
+        public static void GenerateElements<T>(int numberOfElements, GenericArray<T> generatedElements) where T : Human, new()
+        {
+            string FName;
```
Comment (yauhenists) [inline Homework10/Program.cs]: Naming rule - local variables for methods starts with lower case letter

File: Homework10/GenericArray.cs [hunk 7]
```diff
@@ -0,0 +1,51 @@
+﻿namespace Homework10
+{
+    public class GenericArray<T> where T : Human
+    {
+        private T[] _array;
+
+        public GenericArray(int size)
+        {
+            _array = new T[size + 1];
```
Comment (yauhenists) [inline Homework10/GenericArray.cs]: I don't get the point with "size + 1"

File: Homework10/Program.cs [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+            for (int i = 0; i < numberOfElements; i++)
+            {
+                Console.WriteLine("Set first name:");
+                FName = Console.ReadLine();
+                Console.WriteLine("Set last name:");
+                LName = Console.ReadLine();
+                var human = new T();
+                human.FName = FName;
+                human.LName = LName;
+                if (typeof(T) == typeof(Woman))
```
Comment (yauhenists) [inline Homework10/Program.cs]: I think these conditions are not required, you can add element directly

File: Homework10/GenericArray.cs (hunk 5 above)
Comment (yauhenists) [reply Homework10/GenericArray.cs]: Ok, Get the point. Just create one more method for adding element to array. E.g. you have the array [T el1, T el2] and you want to add T el3

**Review by yauhenists — CHANGES_REQUESTED** [review]

File: Homework10/GenericArray.cs (hunk 1 above)
Comment (ritabunets) [reply Homework10/GenericArray.cs]: make sense, updated

File: Homework10/GenericArray.cs (hunk 2 above)
Comment (ritabunets) [reply Homework10/GenericArray.cs]: removed at all

File: Homework10/GenericArray.cs (hunk 3 above)
Comment (ritabunets) [reply Homework10/GenericArray.cs]: updated, but in another way

File: Homework10/GenericArray.cs (hunk 5 above)
Comment (ritabunets) [reply Homework10/GenericArray.cs]: updated

File: Homework10/Program.cs (hunk 6 above)
Comment (ritabunets) [reply Homework10/Program.cs]: oh, yep. fixed

File: Homework10/Program.cs (hunk 8 above)
Comment (ritabunets) [reply Homework10/Program.cs]: right, it works w/o type verification, removed

File: Homework10/GenericArray.cs (hunk 4 above)
Comment (ritabunets) [reply Homework10/GenericArray.cs]: fixed with pain

File: Homework10/GenericArray.cs [hunk 9]
```diff
@@ -1,51 +1,60 @@
...
-            }
-            for (int i = index; i < tmp.Length; i++)
+            if (index < 0 || index >= _array.Length-1)
             {
-                tmp[i] = _array[i + 1];
+                throw new Exception("Set index is out of range, nothing is removed.");
             }
-            _array = tmp;
+            for (int i = index; i < _array.Length - 1; i++)
+                {
```
Comment (yauhenists) [inline Homework10/GenericArray.cs]: Please move code block under "for" loop

**Review by yauhenists — CHANGES_REQUESTED** [review]

**Review by yauhenists — APPROVED** [review]

**Pull request merged**